File: mobile/app/src/main/java/vn/lailua/app/ui/scene/TopVehicles.kt
package vn.lailua.app.ui.scene

import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke

/*
 * Xe nhìn từ trên xuống (chuyển từ sprites.tsx / TopVehicle): mũi xe hướng +x, tâm tại gốc.
 */

private val GLASS = Color(0xFF1F2A3A)
private val SHADOW = Color(0x66000000)

val TOP_LENGTH = mapOf("car" to 40f, "truck" to 56f, "bus" to 64f, "moto" to 26f, "bike" to 22f, "ambulance" to 46f, "fire" to 58f, "police" to 42f)

private fun DrawScope.rr(color: Color, x: Float, y: Float, w: Float, h: Float, r: Float = 0f) =
    drawRoundRect(color, Offset(x, y), Size(w, h), CornerRadius(r, r))

private fun DrawScope.lightBar(x: Float, on: Boolean) {
    rr(Color(0xFF0F172A), x - 3, -7f, 6f, 14f, 2f)
    rr(if (on) Color(0xFFEF4444) else Color(0x55EF4444), x - 2.5f, -6.5f, 5f, 6f, 1.5f)
    rr(if (!on) Color(0xFF3B82F6) else Color(0x553B82F6), x - 2.5f, 0.5f, 5f, 6f, 1.5f)
}

private fun DrawScope.blinker(side: String, len: Float, w: Float, on: Boolean) {
    if (!on) return
    val y = if (side == "left") -w / 2 else w / 2
    drawCircle(Color(0xFFFBBF24), 2.6f, Offset(len / 2 - 2, y))
    drawCircle(Color(0xFFFBBF24), 2.6f, Offset(-len / 2 + 2, y))
}

/**
 * @param blink hướng xi-nhan ("left"/"right"/null)
 * @param tick pha nhấp nháy (đổi mỗi ~0,4 s) cho đèn ưu tiên & xi-nhan
 */
fun DrawScope.drawTopVehicle(kind: String, color: Color, blink: String?, tick: Boolean) {
    when (kind) {
        "car", "police" -> {
            val body = if (kind == "police") Color(0xFFF8FAFC) else color
            rr(SHADOW, -19f, -9f, 42f, 22f, 7f)
            rr(body, -21f, -11f, 42f, 22f, 7f)
            rr(GLASS, 3f, -9f, 8f, 18f, 2f)
            rr(GLASS, -15f, -8.5f, 5f, 17f, 2f)
            rr(body.copy(alpha = 0.85f), -10f, -8f, 13f, 16f, 2f)
            if (kind == "police") {
                rr(Color(0xE61D4ED8), -21f, -3f, 42f, 6f)
                lightBar(-3f, tick)
            }
            drawCircle(Color(0xFFFEF9C3), 1.8f, Offset(19f, -7f))
            drawCircle(Color(0xFFFEF9C3), 1.8f, Offset(19f, 7f))
            if (blink != null) blinker(blink, 42f, 22f, tick)
        }
        "truck" -> {
            rr(SHADOW, -26f, -10f, 56f, 24f, 3f)
            rr(color, -28f, -12f, 42f, 24f, 2f)
            for (y in listOf(-8f, 0f, 8f)) drawLine(Color(0x22000000), Offset(-24f, y), Offset(10f, y))
            rr(Color(0xFFF97316), 15f, -11f, 13f, 22f, 4f)
            rr(GLASS, 22f, -9f, 5f, 18f, 1.5f)
            if (blink != null) blinker(blink, 56f, 24f, tick)
        }
        "bus" -> {
            rr(SHADOW, -30f, -10f, 64f, 24f, 5f)
            rr(color, -32f, -12f, 64f, 24f, 5f)
            rr(GLASS, 25f, -10f, 5f, 20f, 2f)
            rr(Color(0x59FFFFFF), -27f, -8f, 48f, 16f, 3f)
            rr(Color(0x8064748B), -20f, -5f, 8f, 10f, 1.5f)
            rr(Color(0x8064748B), 0f, -5f, 8f, 10f, 1.5f)
            if (blink != null) blinker(blink, 64f, 24f, tick)
        }
        "ambulance" -> {
            rr(SHADOW, -21f, -9f, 46f, 22f, 5f)
            rr(Color.White, -23f, -11f, 46f, 22f, 5f)
            rr(GLASS, 10f, -9f, 7f, 18f, 2f)
            rr(Color(0xFFEF4444), -23f, -11f, 46f, 4f)
            rr(Color(0xFFEF4444), -23f, 7f, 46f, 4f)
            drawLine(Color(0xFFEF4444), Offset(-10f, -6f), Offset(-10f, 6f), strokeWidth = 4f)
            drawLine(Color(0xFFEF4444), Offset(-16f, 0f), Offset(-4f, 0f), strokeWidth = 4f)
            lightBar(4f, tick)
            if (blink != null) blinker(blink, 46f, 22f, tick)
        }
        "fire" -> {
            rr(SHADOW, -27f, -10f, 58f, 24f, 4f)
            rr(Color(0xFFDC2626), -29f, -12f, 58f, 24f, 4f)
            rr(GLASS, 20f, -10f, 6f, 20f, 2f)
            drawLine(Color(0xFFE5E7EB), Offset(-25f, -4f), Offset(12f, -4f), strokeWidth = 1.6f)
            drawLine(Color(0xFFE5E7EB), Offset(-25f, 4f), Offset(12f, 4f), strokeWidth = 1.6f)
            lightBar(14f, tick)
            if (blink != null) blinker(blink, 58f, 24f, tick)
        }
        "moto" -> {
            rr(SHADOW, -11f, -3f, 26f, 8f, 4f)
            rr(Color(0xFF111111), -13f, -3f, 8f, 6f, 3f)
            rr(Color(0xFF111111), 6f, -3f, 8f, 6f, 3f)
            rr(color, -8f, -4f, 18f, 8f, 4f)
            drawCircle(Color(0xFFFACC15), 4f, Offset(-1f, 0f))
            drawLine(Color(0xFF1F2937), Offset(7f, -6f), Offset(7f, 6f), strokeWidth = 2f)
            if (blink != null) blinker(blink, 26f, 10f, tick)
        }
        else -> { // bike
            rr(Color(0xFF111111), -11f, -2f, 7f, 4f, 2f)
            rr(Color(0xFF111111), 4f, -2f, 7f, 4f, 2f)
            drawLine(color, Offset(-8f, 0f), Offset(8f, 0f), strokeWidth = 3f)
            drawCircle(Color(0xFF2563EB), 3.5f, Offset(-1f, 0f))
            drawLine(Color(0xFF1F2937), Offset(6f, -5f), Offset(6f, 5f), strokeWidth = 1.5f)
        }
    }
}

/** Mũi tên ý định (tam giác) */
fun DrawScope.arrowHead(color: Color) {
    val p = Path().apply { moveTo(0f, 0f); lineTo(-12f, -7f); lineTo(-12f, 7f); close() }
    drawPath(p, color)
}

fun DrawScope.ring(radius: Float, color: Color, width: Float, dashed: Boolean = false) {
    drawCircle(
        color, radius, Offset.Zero,
        style = Stroke(width, pathEffect = if (dashed) androidx.compose.ui.graphics.PathEffect.dashPathEffect(floatArrayOf(6f, 5f)) else null),
    )
}

File: ios/LaiLua/UI/Scene/TopVehicles.swift
import SwiftUI

/*
 * Xe nhìn từ trên xuống (chuyển từ sprites.tsx / TopVehicle): mũi xe hướng +x, tâm tại gốc.
 */

private let GLASS = Color(hex: 0x1F2A3A)
private let SHADOW = Color.black.opacity(0.4)

let TOP_LENGTH: [String: CGFloat] = ["car": 40, "truck": 56, "bus": 64, "moto": 26, "bike": 22, "ambulance": 46, "fire": 58, "police": 42]

private func lightBar(_ c: GraphicsContext, _ x: CGFloat, on: Bool) {
    c.rr(Color(hex: 0x0F172A), x - 3, -7, 6, 14, 2)
    c.rr(on ? Color(hex: 0xEF4444) : Color(hex: 0xEF4444, alpha: 0.33), x - 2.5, -6.5, 5, 6, 1.5)
    c.rr(!on ? Color(hex: 0x3B82F6) : Color(hex: 0x3B82F6, alpha: 0.33), x - 2.5, 0.5, 5, 6, 1.5)
}

private func blinker(_ c: GraphicsContext, side: String, len: CGFloat, w: CGFloat, on: Bool) {
    guard on else { return }
    let y = side == "left" ? -w / 2 : w / 2
    c.circle(Color(hex: 0xFBBF24), 2.6, at: P(len / 2 - 2, y))
    c.circle(Color(hex: 0xFBBF24), 2.6, at: P(-len / 2 + 2, y))
}

/// - Parameters:
///   - blink: hướng xi-nhan ("left"/"right"/nil)
///   - tick: pha nhấp nháy (đổi mỗi ~0,4 s) cho đèn ưu tiên & xi-nhan
func drawTopVehicle(_ c: GraphicsContext, kind: String, color: Color, blink: String?, tick: Bool) {
    switch kind {
    case "car", "police":
        let body = kind == "police" ? Color(hex: 0xF8FAFC) : color
        c.rr(SHADOW, -19, -9, 42, 22, 7)
        c.rr(body, -21, -11, 42, 22, 7)
        c.rr(GLASS, 3, -9, 8, 18, 2)
        c.rr(GLASS, -15, -8.5, 5, 17, 2)
        c.rr(body.opacity(0.85), -10, -8, 13, 16, 2)
        if kind == "police" {
            c.rr(Color(hex: 0x1D4ED8, alpha: 0.9), -21, -3, 42, 6)
            lightBar(c, -3, on: tick)
        }
        c.circle(Color(hex: 0xFEF9C3), 1.8, at: P(19, -7))
        c.circle(Color(hex: 0xFEF9C3), 1.8, at: P(19, 7))
        if let blink { blinker(c, side: blink, len: 42, w: 22, on: tick) }
    case "truck":
        c.rr(SHADOW, -26, -10, 56, 24, 3)
        c.rr(color, -28, -12, 42, 24, 2)
        for y: CGFloat in [-8, 0, 8] { c.line(Color.black.opacity(0.13), P(-24, y), P(10, y)) }
        c.rr(Color(hex: 0xF97316), 15, -11, 13, 22, 4)
        c.rr(GLASS, 22, -9, 5, 18, 1.5)
        if let blink { blinker(c, side: blink, len: 56, w: 24, on: tick) }
    case "bus":
        c.rr(SHADOW, -30, -10, 64, 24, 5)
        c.rr(color, -32, -12, 64, 24, 5)
        c.rr(GLASS, 25, -10, 5, 20, 2)
        c.rr(Color.white.opacity(0.35), -27, -8, 48, 16, 3)
        c.rr(Color(hex: 0x64748B, alpha: 0.5), -20, -5, 8, 10, 1.5)
        c.rr(Color(hex: 0x64748B, alpha: 0.5), 0, -5, 8, 10, 1.5)
        if let blink { blinker(c, side: blink, len: 64, w: 24, on: tick) }
    case "ambulance":
        c.rr(SHADOW, -21, -9, 46, 22, 5)
        c.rr(.white, -23, -11, 46, 22, 5)
        c.rr(GLASS, 10, -9, 7, 18, 2)
        c.rr(Color(hex: 0xEF4444), -23, -11, 46, 4)
        c.rr(Color(hex: 0xEF4444), -23, 7, 46, 4)
        c.line(Color(hex: 0xEF4444), P(-10, -6), P(-10, 6), width: 4)
        c.line(Color(hex: 0xEF4444), P(-16, 0), P(-4, 0), width: 4)
        lightBar(c, 4, on: tick)
        if let blink { blinker(c, side: blink, len: 46, w: 22, on: tick) }
    case "fire":
        c.rr(SHADOW, -27, -10, 58, 24, 4)
        c.rr(Color(hex: 0xDC2626), -29, -12, 58, 24, 4)
        c.rr(GLASS, 20, -10, 6, 20, 2)
        c.line(Color(hex: 0xE5E7EB), P(-25, -4), P(12, -4), width: 1.6)
        c.line(Color(hex: 0xE5E7EB), P(-25, 4), P(12, 4), width: 1.6)
        lightBar(c, 14, on: tick)
        if let blink { blinker(c, side: blink, len: 58, w: 24, on: tick) }
    case "moto":
        c.rr(SHADOW, -11, -3, 26, 8, 4)
        c.rr(Color(hex: 0x111111), -13, -3, 8, 6, 3)
        c.rr(Color(hex: 0x111111), 6, -3, 8, 6, 3)
        c.rr(color, -8, -4, 18, 8, 4)
        c.circle(Color(hex: 0xFACC15), 4, at: P(-1, 0))
        c.line(Color(hex: 0x1F2937), P(7, -6), P(7, 6), width: 2)
        if let blink { blinker(c, side: blink, len: 26, w: 10, on: tick) }
    default: // bike
        c.rr(Color(hex: 0x111111), -11, -2, 7, 4, 2)
        c.rr(Color(hex: 0x111111), 4, -2, 7, 4, 2)
        c.line(color, P(-8, 0), P(8, 0), width: 3)
        c.circle(Color(hex: 0x2563EB), 3.5, at: P(-1, 0))
        c.line(Color(hex: 0x1F2937), P(6, -5), P(6, 5), width: 1.5)
    }
}

/// Mũi tên ý định (tam giác)
func arrowHead(_ c: GraphicsContext, _ color: Color) {
    c.fillPath(polygon([P(0, 0), P(-12, -7), P(-12, 7)]), color)
}
